File: sdk/c#/PicarioXPO.RenderAPI.V2/DesignKeyExtractors/EntityNameDesignKey.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Web;

namespace PicarioXPO.RenderAPI.V2.DesignKeyExtractors
{
    internal class EntityNameDesignKey : DesignKey
    {
        public override string GetValues(IEnumerable<XpoUrlObject> designs)
        {
            var xpoUrlObjects = designs as XpoUrlObject[] ?? designs.ToArray();
            var max = xpoUrlObjects.GetMaxObjectNumber();
            
            for (var i = 0; i <= max; i++)
            {
                var urlObject = xpoUrlObjects.FirstOrDefault(x => x.Index == i);
                if (urlObject != null && !string.IsNullOrEmpty(urlObject.Design.EntityName))
                    AddToList(ConvertToBase64UrlString(urlObject.Design.EntityName));
                else if (i != max)
                    AddEmpty();
            }

            if (IsEmpty()) return "";

            return "p.tn=" + GetUrlValue();
        }

        private static string ConvertToBase64UrlString(string value)
        {
            return HttpUtility.UrlEncode(Convert.ToBase64String(Encoding.UTF8.GetBytes(value)));
        }
    }
}

File: sdk/js/PicarioXPO.RenderAPI.V2/XpoUrlGenerator/EntityNameDesignKey.ts
class EntityNameDesignKey extends DesignKey {

    getValues(designs: Array<XpoUrlObject>) {
        var max = UrlGeneratorModule.getMaxObjectNumber(designs);

        for (var i = 0; i <= max; i++) {
            var index = designs.map(e => e.getIndex()).lastIndexOf(i);

            var urlObject = designs[index];
            if (urlObject != null)
                this.addToList(this.convertToBase64UrlString(urlObject.getDesign().getEntityName()));
            else if (i != max)
                this.addEmpty();
        }

        if (this.isEmpty()) return "";

        return "p.tn=" + this.getUrlValue();
    }

    convertToBase64UrlString(value: string) {
        var base64 = new Base64();
        
        return encodeURIComponent(base64.encode(value));
    }
} 